performance/class-mime-types-caching.php

```diff
@@ -0,0 +1,160 @@
+	 * Class initialization.
+	 */
+	public static function init() {
+		global $wp_version;
+
+		if ( isset( $wp_version ) &&
+			version_compare( $wp_version, self::MINIMUM_WORDPRESS_VERSION, '>=' ) ) {
+			static::enable_post_mime_types_caching();
+		}
+	}
+
+	/**
+	 * Enable MIME type caching.
+	 */
+	private static function enable_post_mime_types_caching() {
+		add_filter( 'pre_get_available_post_mime_types', array( __CLASS__, 'get_cached_post_mime_types' ), 10, 2 );
+		add_action( 'add_attachment', array( __CLASS__, 'add_mime_type_to_cache' ), 10, 1 );
+		add_action( 'attachment_updated', array( __CLASS__, 'update_mime_types_cache' ), 10, 3 );
+		add_action( 'delete_attachment', array( __CLASS__, 'delete_mime_types_cache' ), 10 );
+	}
+
+	/**
+	 * Get cached results for get_available_post_mime_types() to avoid a query on every page load.
+	 *
+	 * @param string[]|null $filtered_mime_types An array of MIME types. Default null.
+	 * @param string        $type                The post type name.
+	 * @return array An array of MIME types.
+	 */
+	public static function get_cached_post_mime_types( $filtered_mime_types, $type ) {
+		if ( 'attachment' !== $type ) {
+			return $filtered_mime_types;
+		}
+
+		$mime_types = wp_cache_get( self::MIME_TYPES_CACHE_KEY, self::CACHE_GROUP );
+
+		if ( false === $mime_types ) {
+
+			/**
+			 * Filters the max number of posts to query for dynamic MIME type caching.
+			 *
+			 * @param bool $max_posts_to_query Max number of posts to query.
+			 */
+			$max_posts_to_query = apply_filters( 'vip_max_posts_to_query_for_mime_type_caching', self::MAX_POSTS_TO_QUERY_DEFAULT );
+
+			global $wpdb;
+			$attachment_count = $max_posts_to_query > 0 ? $wpdb->get_var( $wpdb->prepare( "SELECT COUNT(1) FROM {$wpdb->posts} WHERE post_type = %s", $type ) ) : 1;
+			$use_defaults     = $attachment_count > $max_posts_to_query;
+
+			if ( $use_defaults ) {
+				// If there are too many posts to query, use the default MIME types.
+				$available_types = static::get_default_mime_types();
+			} else {
+				// Otherwise, use the same query from core.
+				$available_types = $wpdb->get_col( $wpdb->prepare( "SELECT DISTINCT post_mime_type FROM $wpdb->posts WHERE post_type = %s", $type ) );
+			}
+
+			$mime_types = array(
+				'available_types' => $available_types,
+				'using_defaults'  => $use_defaults,
+			);
+
+			wp_cache_set( self::MIME_TYPES_CACHE_KEY, $mime_types, self::CACHE_GROUP );
+		} else {
+			$available_types = $mime_types['available_types'] ?? null;
+		}
+
+		// If there were any previous MIME types, merge them with the cached MIME types.
+		if ( is_array( $filtered_mime_types ) ) {
+			$available_types = is_array( $available_types ) ? array_unique( array_merge( $filtered_mime_types, $available_types ) ) : $filtered_mime_types;
+		}
+
+		return $available_types;
+	}
+
+	/**
+	 * Get the default MIME types.
+	 *
+	 * @return string[] An array of default MIME types.
+	 */
+	public static function get_default_mime_types() {
```

Comment:
I have performance concerns about this method. `get_post_mime_types()` is a slow function, and this method is essentially partially undoing its work.  So first, get_post_mime_types gets flat mime types and runs regex over available mime types to filter some out and group others by arbitrary type, and then we come in and do the flattening.

Consider this on 10k iterations:
> get_post_mime_types Elapsed: 1.9381890296936

> wp_get_mime_types: 0.00063705444335938


Furthermore, the functional approach is almost 3 times slower than simplistic explode/join on the same 10k iterations:

```php
add_action( 'init', function() {
	$pmt =  array_keys( get_post_mime_types() );
	$i = 0;
	$m = microtime( true );
	while ( $i++ < 10000 ) {
		$functional = array_reduce(
			$pmt,
			function ( $carry, $mime_type ) {
				return array_merge( $carry, explode( ',', $mime_type ) );
			},
			array()
		);
	}
	echo "array_reduce Elapsed: " . microtime( true ) - $m;


	$i = 0;
	$m = microtime( true );
	while ( $i++ < 10000 ) {
		$explode_join = explode( ',', join( ',', $pmt ) );
	}

	echo "explode/join elapsed: " . microtime(true) - $m;
} );
```

> array_reduce Elapsed: 0.0083451271057129
> explode/join elapsed: 0.003554105758667

Reply:
Thanks for the feedback, @rinatkhaziev. The join/explode approach seems to be more performant indeed. Code updated.

In regards to the `get_post_mime_types()` call, I utilized it because the grid view populates the media type dropdown based on the results from calling that function too, and I wanted the list view to be in parity to ensure consistency in the future:
https://github.com/WordPress/WordPress/blob/6.4.1/wp-includes/media.php#L4702

In our current version of `get_default_mime_types()`, we end up with an array of 26 items (in a vanilla WP install) after manipulating the results from `get_post_mime_types()`. While doing `array_values( wp_get_mime_types() )` in this function would be quicker, it returns 96 items corresponding to the unfiltered data.

If used as is, the end result in the dropdown will be similar, but there are at least two side effects:
1. More than triple the data would be cached/read every time (3.68KB vs 1.07KB when serialized)
2. Having more items would mean the caller functions at the upper levels (like `wp_match_mime_types()` coming from `media_upload_library_form()`) would need to perform more iterations.

Reply:
I clearly lack context here :) Will this method only be called in wp-admin?

Reply:
Correct, only when visiting the list view in the media library on a site with a large number of attachments and if the results are not already cached.